crates/lib/src/qpu/api.rs

```diff
@@ -144,21 +171,152 @@ pub async fn retrieve_results(
+    async fn get_controller_client(
+        &self,
+        client: &Qcs,
+        connection_strategy: ConnectionStrategy,
+    ) -> Result<ControllerClient<RefreshService<Channel, ClientConfiguration>>, QpuApiError> {
+        match self {
+            Self::EndpointId(endpoint_id) => {
+                let endpoint = get_endpoint(&client.get_openapi_client(), endpoint_id).await?;
+                let grpc_address = endpoint
+                    .addresses
+                    .grpc
+                    .ok_or_else(|| QpuApiError::EndpointNotFound(endpoint_id.into()))?;
+                Self::grpc_address_to_client(&grpc_address, client)
+            }
+            Self::QuantumProcessorId(quantum_processor_id) => {
+                let address = match connection_strategy {
+                    ConnectionStrategy::GatewayAlways => {
+                        self.get_gateway_address(quantum_processor_id, client)
+                            .await?
+                    }
+                    ConnectionStrategy::DirectAccessAlways => {
+                        self.get_default_endpoint_address(quantum_processor_id, client)
+                            .await?
+                    }
+                };
+                Self::grpc_address_to_client(&address, client)
+            }
+        }
+    }
+
+    fn grpc_address_to_client(
+        address: &str,
+        client: &Qcs,
+    ) -> Result<ControllerClient<RefreshService<Channel, ClientConfiguration>>, QpuApiError> {
+        let uri = parse_uri(address).map_err(QpuApiError::GrpcError)?;
+        let channel = get_channel(uri).map_err(|err| QpuApiError::GrpcError(err.into()))?;
+        let service = wrap_channel_with(channel, client.get_config().clone());
+        Ok(ControllerClient::new(service))
+    }
+
+    async fn get_gateway_address(
+        &self,
+        quantum_processor_id: &str,
+        client: &Qcs,
+    ) -> Result<String, QpuApiError> {
+        let mut gateways = Vec::new();
+        let mut next_page_token = None;
+        loop {
+            let accessors = list_quantum_processor_accessors(
+                &client.get_openapi_client(),
+                quantum_processor_id,
+                Some(100),
+                next_page_token.as_deref(),
+            )
+            .await?;
+            gateways.extend(accessors.accessors.into_iter().filter(|acc| {
+                acc.live
+                    // `as_deref` needed to work around the `Option<Box<_>>` type.
+                    && acc.access_type.as_deref() == Some(&QuantumProcessorAccessorType::GatewayV1)
+            }));
+            next_page_token = accessors.next_page_token.clone();
+            if next_page_token.is_none() {
+                break;
+            }
+        }
+        gateways.sort_by_key(|acc| acc.rank);
+        gateways
+            .first()
+            .map(|accessor| accessor.url.clone())
+            .ok_or_else(|| QpuApiError::GatewayNotFound(quantum_processor_id.to_string()))
```

Comment:
FYI, `min_by_key` is a little more efficient than `sort` -> `first`.

As an optimization, we can avoid allocating the `Vec` if we just accumulate the minimum value as we loop. In practice, not likely to affect anything (this is called rarely and is dominated by IO time), but also an easy fix.

Reply:
This was lifted out of the old client code, but I went ahead and refactored it.

Just to clarify, `accessor.rank` is an `Option`, meaning it could be `None`. How should that value rank against accessors with `Some(i64)`. By default Rust rules it's the minimum value (`None < Some(0)`). I have a feeling we want the opposite though, so I defaulted them to i64::MAX. Is that correct? Or does `None` signify the preferred gateway?

Reply:
Yep, I agree, good spot - no rank is "lowest rank"